activation/handler_v2.go

```diff
@@ -0,0 +1,177 @@
+	"fmt"
+	"time"
+
+	"github.com/spacemeshos/post/shared"
+	"go.uber.org/zap"
+
+	"github.com/spacemeshos/go-spacemesh/activation/wire"
+	"github.com/spacemeshos/go-spacemesh/atxsdata"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/datastore"
+	"github.com/spacemeshos/go-spacemesh/log"
+	mwire "github.com/spacemeshos/go-spacemesh/malfeasance/wire"
+	"github.com/spacemeshos/go-spacemesh/p2p"
+	"github.com/spacemeshos/go-spacemesh/signing"
+	"github.com/spacemeshos/go-spacemesh/sql/atxs"
+	"github.com/spacemeshos/go-spacemesh/system"
+)
+
+//go:generate mockgen -typed -source=./handler_v2.go -destination=mocks_handler_v2.go -package=activation
+
+type nipostValidatorV2 interface {
+	IsVerifyingFullPost() bool
+	VRFNonceV2(smesherID types.NodeID, commitment types.ATXID, vrfNonce types.VRFPostIndex, numUnits uint32) error
+	PostV2(
+		ctx context.Context,
+		smesherID types.NodeID,
+		commitment types.ATXID,
+		post *wire.PostV1,
+		challenge []byte,
+		numUnits uint32,
+		opts ...validatorOption,
+	) error
+}
+
+type HandlerV2 struct {
+	local           p2p.Peer
+	cdb             *datastore.CachedDB
+	atxsdata        *atxsdata.Data
+	edVerifier      *signing.EdVerifier
+	clock           layerClock
+	tickSize        uint64
+	goldenATXID     types.ATXID
+	nipostValidator nipostValidatorV2
+	beacon          AtxReceiver
+	tortoise        system.Tortoise
+	log             *zap.Logger
+	fetcher         system.Fetcher
+}
+
+func (h *HandlerV2) processATX(
+	ctx context.Context,
+	peer p2p.Peer,
+	watx *wire.ActivationTxV2,
+	blob []byte,
+	received time.Time,
+) (*mwire.MalfeasanceProof, error) {
+	exists, err := atxs.Has(h.cdb, watx.ID())
+	if err != nil {
+		return nil, fmt.Errorf("failed to check if atx exists: %w", err)
+	}
+	if exists {
+		return nil, nil
+	}
+
+	h.log.Debug(
+		"processing atx",
+		log.ZContext(ctx),
+		zap.Stringer("atx_id", watx.ID()),
+		zap.Uint32("publish", watx.PublishEpoch.Uint32()),
+		zap.Stringer("smesherID", watx.Smesher()),
+	)
+
+	if err := h.syntacticallyValidate(ctx, watx); err != nil {
+		return nil, fmt.Errorf("atx %s syntactically invalid: %w", watx.ID(), err)
+	}
+
+	// TODO:
+	// 1. fetch dependencies
+	// 2. syntactically validate dependencies
+	// 3. contextually validate ATX
+	// 4. store ATX
+	return nil, nil
+}
+
+// Syntactically validate an ATX.
+// TODOs:
+// 1. support marriages
+// 2. support merged ATXs.
+func (h *HandlerV2) syntacticallyValidate(ctx context.Context, atx *wire.ActivationTxV2) error {
+	if !h.edVerifier.Verify(signing.ATX, atx.Smesher(), atx.SignedBytes(), atx.Signature) {
+		return fmt.Errorf("invalid atx signature: %w", errMalformedData)
+	}
+	if atx.PositioningATX == types.EmptyATXID {
+		return errors.New("empty positioning atx")
+	}
+	// TODO: support marriages
+	if len(atx.Marriages) != 0 {
+		return errors.New("marriages are not supported")
+	}
+
+	current := h.clock.CurrentLayer().GetEpoch()
+	if atx.PublishEpoch > current+1 {
```

Comment:
I know we do the same thing for ATX v1, but I'm not sure we should.

Let's assume the clock on the node verifying an ATX runs slow - lets say it's even one full day behind. Then on the first day of epoch `n` it would refuse ATXs that have been published on the first day of the epoch, because it thinks it is still in epoch `n-1`. I think this is acceptable for a node with such a huge delay in its clock, because the opposite is worse:

On the first day of epoch `n` any node (even ones with correct calibrated clocks) now consider an ATX published for epoch `n+1` (2 weeks into the future) as valid.

Reply:
I'm not 100% sure what the motivation behind this check was. If the motivation is to accept ATXs received on the publish epoch boundary, then perhaps the check could be changed to verify that we received the ATX within the allowed window. The ATX would be valid if `receivedTime > atx.PublishEpoch.StartTime().Sub(<acceptable drift>)`, where acceptable drift is something like 1 hour. Wdyt?

Reply:
Yeah maybe - for the exact value we should maybe accept the delay that we use for grading - `ATXGradeDelay` or some value relative to it. I think any ATX that is received less than 4 x that value before the end of the epoch is possibly considered too late.

Reply:
Theoretically we don't need to check publish epoch at all. The current check just serves to prevent someone from spamming the network until we have fixed https://github.com/spacemeshos/go-spacemesh/issues/4756